content/posts/2021/test-python-functions/index.md

```diff
@@ -0,0 +1,387 @@
+
+<!-- more -->
+## Intro
+
+For this demo we are going to create a tiny calculator function. This gives us just enough code to demonstrate interesting but very common test cases:
+
+1. is the request correct, .e.g did the request specify a valid math operation and two numbers,
+2. is the response correct, i.e. the calculation.
+
+Almost all functions will want these kind of tests and, as we will see, they are easy to write.
+
+If you want to skip to the final result, checkout the sample repo here: https://github.com/LucasRoesler/pytest-openfaas-sample
+
+### Start the project
+```sh
+$ mkdir pytest-sample
+$ cd pytest-sanmple
+$ faas-cli template store pull python3-flask
+$ mv calc.yml stack.yml
+```
+
+**Note** Don't forget to add this section to your stack file so that the `faas-cli` will know how to automatically pull the template
+
+```yaml
+configuration:
+  templates:
+    - name: python3-flask
+      source: https://github.com/openfaas/python-flask-template
+
+```
+
+### Setup the local python environment
+I use [conda](https://docs.conda.io/projects/conda/en/latest/) for my local virtual environments, but you can of course use [`virtualenv`](https://virtualenv.readthedocs.org/en/latest/) or [`venv`](https://docs.python.org/3/library/venv.html) (see also [RealPython's primer on virtual environments](https://realpython.com/python-virtual-environments-a-primer/)).
+
+In short this creates an isolated and repeatable development environment which you can delete and recreate if you need.
+
+```sh
+$ conda create -n pytest-sample tox
+$ conda activate pytest-sample
+$ cat <<EOF >> requirements.txt
+pydantic==1.7.3
+flask==1.1.2
+EOF
+$ cat <<EOF >> dev.txt
+tox
+pytest
+black
+pylint
+EOF
+$ conda install --yes --file requirements.txt
+$ conda install --yes --file dev.txt
+```
+
+Now we are ready to develop the calculator.
+
+### Add the first tests
+Let's start with some simple tests:
+
+```sh
+$ touch calc/handler_test.py
+```
+
+then add the following test cases for a couple of our happy paths
+
+```py
+import calc.handler as h
+
+
+class TestParsing:
+    def test_operation_addition(self):
+        req = '{"op": "+", "var1": "1.0", "var2": 0}'
+        resp, code = h.handle(req)
+        assert code == 200
+        assert resp["value"] == 1.0
+
+    def test_operation_multiplication(self):
+        req = '{"op": "*", "var1": "100.01", "var2": 1}'
+        resp, code = h.handle(req)
+        assert code == 200
+        assert resp["value"] == 100.01
+```
+
+At this point, we have followed normal conventions for pytest, by default, pytest will look for files that match the pattern `*_test.py` and then the run the functions that  match `test_*` (including matching methods if the class is named `Test*`, which is what we did above).
```

Comment:
either `as we have followed normal conventions for pytest, by default pytest will look` (removing the comma after the default)

or `we have followed normal conventions for pytest, so pytest will by default`